src/librustc/hir/lowering.rs

```diff
@@ -2146,10 +2147,11 @@ impl<'a> LoweringContext<'a> {
                 //     let result = match ::std::iter::IntoIterator::into_iter(<head>) {
                 //       mut iter => {
                 //         [opt_ident]: loop {
-                //           match ::std::iter::Iterator::next(&mut iter) {
-                //             ::std::option::Option::Some(<pat>) => <body>,
+                //           let <pat> = match ::std::iter::Iterator::next(&mut iter) {
```

Comment:
This changes semantics, although I can't think of any situation in which this would fail.
FWIW when I asked about `while let Some(x) = iter.next()` - that would correspond to the old desugaring, not your new one, wouldn't it?